### PR TITLE
Add reach whitepaper

### PR DESCRIPTION
**Pull request by hidayetaksu**
Add the whitepaper on reach implementation best practices.

### CODE REVIEW

**Comment by hidayetaksu** [comment]
This PR adds a white-paper on implementing reach measurement within the Privacy Sandbox, via the Shared Storage and Private Aggregation APIs.

**Comment by michalkalisz** [comment]
Thank you for this valuable document. It is not only insightful in the context of reach measurement but also has broader implications for aggregated reporting.

I am curious if you have considered the possibility of interleaving between checking if a report has been sent and marking it as sent. 
```javascript
    const hasReportedContent = (await sharedStorage.get(key)) === 'true';
    (...)
    await sharedStorage.set(key, true);
```

It seems to me that the ReachMeasurementOperation is not synchronized and may be executed simultaneously on two different worklets, which is quite realistic in the case of impressions (when multiple ads from the same advertiser are displayed on a single page).

File: reach-whitepaper.md [hunk 1]
```diff
@@ -0,0 +1,1019 @@
...
+In particular, we assume that all queries correspond to the same time periods. 
+However, we may want to measure time-related queries. 
+A popular family of such queries is cumulative queries, which are queries of 
+the form “what is the total reach up until timestep `t`?” where 
+`t` varies from `1, …, T`.
+For example, one might want to measure the reach up until today everyday for the
+ entire month. In this case, we would have `T = 30` and “timestep” being a day.
+
+_Note: Both methods described in this section are limited by the 
+[retention policy](https://github.com/WICG/shared-storage?tab=readme-ov-file#data-retention-policy) of Shared Storage which is 30 days from last write._
```
Comment (alexmturner) [inline reach-whitepaper.md:393]: hyper nit: could remove the `?tab=readme-ov-file`

File: reach-whitepaper.md [hunk 2]
```diff
@@ -0,0 +1,1019 @@
...
+Specifically, as we will see below, we need to set the capping parameter to be
+small to ensure the noise is small but doing so results in a large error due to
+capping. 
+This tension means that we have to be very careful in terms of parameter 
+settings to ensure (reasonably) accurate estimates.
+
+The overall scheme of the reach estimation involving sketches is described on 
+the following diagram.
+
+![alt_text](figs/diagram.png "computation diagram")
```
Comment (alexmturner) [inline reach-whitepaper.md:572]: can we add a slightly more descriptive alt text?

File: reach-whitepaper.md [hunk 3]
```diff
@@ -0,0 +1,1019 @@
...
+    Level
+<p>
+
+    (granularity)
+   </td>
+   <td style="background-color: #cccccc">
+    Query dimensions
+   </td>
+   <td style="background-color: #cccccc">
+    Power law shape (b) param \
```
Comment (alexmturner) [inline reach-whitepaper.md:832]: the backslash doesn't appear to be doing the right new line behavior

File: reach-whitepaper.md [hunk 4]
```diff
@@ -0,0 +1,1019 @@
...
+Listing 1: pseudocode to generate synthetic data. 
+
+
+<table>
+  <tr>
+   <td style="background-color: #cccccc">
+    Level
+<p>
+
+    (granularity)
```
Comment (alexmturner) [inline reach-whitepaper.md:826]: This appears to have been treated as a code block

File: reach-whitepaper.md [hunk 5]
```diff
@@ -0,0 +1,1019 @@
...
+In accordance with the pseudocode presented in Listing 1, synthetic datasets of 
+variable sizes, capping values, and time windows were generated for experimental
+evaluation.
+
+
+### Error Metrics
+
+In our experiments we are going to use $`\text{RMSRE}_\tau`$ metric defined as 
+[follows](https://developers.google.com/privacy-sandbox/relevance/attribution-reporting/design-decisions#expandable-9):
+$`
```
Comment (alexmturner) [inline reach-whitepaper.md:882]: Let's maybe use `$$` here to make it a block

**Review by alexmturner — COMMENTED** [review]
Thanks! Generally looks good -- just a few nits :)

File: reach-whitepaper.md [hunk 6]
```diff
@@ -0,0 +1,1019 @@
...
+use `L1_BUDGET`.
+See the
+[“Noise and scaling” section](https://developers.google.com/privacy-sandbox/relevance/private-aggregation/fundamentals#noise_and_scaling) 
+of the Private Aggregation fundamentals article to learn more.
+
+
+### Single Reach Measurement with Capping
+
+An implicit assumption in the previous section is that each 
+[privacy unit](https://www.google.com/url?q=https://github.com/patcg-individual-drafts/private-aggregation-api?tab%3Dreadme-ov-file%23implementation-plan&sa=D&source=docs&ust=1715709807432613&usg=AOvVaw2F52S_hQoFFfN3FjYLRq18)
```
[the remark this answers is not in the document]
Comment (hidayetaksu) [reply reach-whitepaper.md]: Done.

File: reach-whitepaper.md [hunk 7]
```diff
@@ -0,0 +1,1019 @@
...
+experimental errors in our dataset. 
+However, it is plausible that other types of
+sketches can allow for better utility.
+
+As the sketch is simply a histogram, it is simple to implement via the Shared
+Storage and Private Aggregation APIs. 
+Again, for simplicity, we only include the code snippet for the cap-one case
+below. (Note that if there are many different content ids, one might consider 
+splitting them into groups using 
+[filtering ID’s](https://www.google.com/url?q=https://github.com/patcg-individual-drafts/private-aggregation-api/blob/main/flexible_filtering.md&sa=D&source=docs&ust=1717257234668602&usg=AOvVaw01fIhnBOOZisrNsyNBMHqm).)
```
[the remark this answers is not in the document]
Comment (hidayetaksu) [reply reach-whitepaper.md]: Done.

File: reach-whitepaper.md [hunk 8]
```diff
@@ -0,0 +1,1019 @@
...
+corresponding curve fitting.
+
+Furthermore, we examined the impact of the time window on the reach. Reach does
+not demonstrate a linear relationship with time window size.
+Instead, it exhibits a diminishing return compared to linear growth as
+illustrated in Figure 3. 
+This discount follows a power-law function relative to the size of the time
+window.
+
+![Reach estimates receive a discount as a function of cap value.](figs/cap-discount.png)
```
[the remark this answers is not in the document]
Comment (hidayetaksu) [reply reach-whitepaper.md]: renamed figd to reach_whitepaper_figs. it renders correctly on review.

File: reach-whitepaper.md [hunk 9]
```diff
@@ -0,0 +1,1019 @@
+# Reach Implementation Best Practices in the Privacy Sandbox Shared Storage + Private Aggregation APIs
```
[the remark this answers is not in the document]
Comment (hidayetaksu) [reply reach-whitepaper.md]: Done.

File: reach-whitepaper.md (hunk 5 above)
Comment (hidayetaksu) [reply reach-whitepaper.md]: I tried, but apparently, GitHub doesn't render LaTeX math expressions (using $$ delimiters) correctly with the current code. Kept $ as it just works.

File: reach-whitepaper.md (hunk 4 above)
Comment (hidayetaksu) [reply reach-whitepaper.md]: right. replacing <p> with <br> solved it.

File: reach-whitepaper.md (hunk 2 above)
Comment (hidayetaksu) [reply reach-whitepaper.md]: Added a more descriptive alt text.

File: reach-whitepaper.md (hunk 1 above)
Comment (hidayetaksu) [reply reach-whitepaper.md]: Done.

File: reach-whitepaper.md [hunk 10]
```diff
@@ -0,0 +1,1019 @@
...
+![error comparison](figs/errors_comparison.png)
+
+We are hopeful that this paper will enable advertising technologists to
+comprehend the trade-offs associated with various methodologies and develop a
+structured path to implementing reach using the APIs.
+
+
+## Acknowledgements
+
+We would like to thank Badih Ghazi, Charlie Harrison, Ravi Kumar, Asha Menon, and Alex Turner for their input and help with this work.
```
[the remark this answers is not in the document]
Comment (hidayetaksu) [reply reach-whitepaper.md]: added one. Thanks for suggestion.

File: reach-whitepaper.md (hunk 5 above)
Comment (alexmturner) [reply reach-whitepaper.md]: I think it's meant to just use `$$` without any backticks, but that doesn't seem to be working for me either. Maybe let's use the code block syntax (using the 'language' of `math`) instead? (Second part of https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/writing-mathematical-expressions#writing-expressions-as-blocks.) That seems to be working for me.

File: reach-whitepaper.md (hunk 2 above)
Comment (alexmturner) [reply reach-whitepaper.md]: Ah sorry I meant more could we replace `alt_text` with something like `computation diagram` (although the longer title is good too!)

File: reach-whitepaper.md (hunk 5 above)
Comment (hidayetaksu) [reply reach-whitepaper.md]: code block syntax works great. used it.

File: reach-whitepaper.md (hunk 2 above)
Comment (hidayetaksu) [reply reach-whitepaper.md]: Thanks for clarification. I used a more longer title and moved description in alt text.

**Review by alexmturner — APPROVED** [review]
Thanks! LGTM

**Comment by alexmturner** [comment]
@michalkalisz Thanks for your feedback! Definitely hear the concern and agree this could be an issue. We're looking into the options here, but one possibility could be extending the [Web Locks API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Locks_API) to function within Shared Storage as well. As a mitigation for now, you might be able to track Shared Storage calls on the page and wait for some amount of time between calls to sharedStorage.run(). (This won't completely fix the problem, but might make it rarer.)

Also, wanted to let you know that we've filed issue #153 for tracking feedback on the whitepaper generally in case you had more.

**Pull request merged**